File: handlers/group.py
import logging
from aiogram import Router, F, Bot
from aiogram.types import Message, ChatMemberUpdated, InlineKeyboardMarkup, InlineKeyboardButton, CallbackQuery
from aiogram.filters import ChatMemberUpdatedFilter, IS_MEMBER, IS_NOT_MEMBER, ADMINISTRATOR
from database.sqlite_db import SQLiteDatabase

router = Router()
logger = logging.getLogger(__name__)

def get_admin_check_keyboard(bot_member, is_forum: bool = True):
    is_admin = bot_member.status == "administrator"
    can_manage_topics = getattr(bot_member, 'can_manage_topics', False)
    can_pin_messages = getattr(bot_member, 'can_pin_messages', False)

    keyboard = []

    if not (is_admin and can_manage_topics and can_pin_messages):
        if is_admin:
            text = "Permissions missing ⚠️"
            cb_data = "admin_missing_perms"
        else:
            text = "Bot is not Admin 🚫"
            cb_data = "admin_not_admin"
        keyboard.append([InlineKeyboardButton(text=text, callback_data=cb_data)])

    if not is_forum:
        keyboard.append([InlineKeyboardButton(text="Topics not enabled 🚫", callback_data="forum_not_enabled")])

    if not keyboard:
        return None

    return InlineKeyboardMarkup(inline_keyboard=keyboard)

@router.my_chat_member(ChatMemberUpdatedFilter(member_status_changed=IS_NOT_MEMBER >> IS_MEMBER))
async def bot_added_to_group(event: ChatMemberUpdated, bot: Bot, db: SQLiteDatabase):
    if event.chat.type not in ["group", "supergroup"]:
        return

    if event.chat.type != "supergroup":
        await bot.send_message(event.chat.id, "this group is not a supergroup. make it supergroup by enabling topics then add me again")
        await bot.leave_chat(event.chat.id)
        return

    owner_id = await db.get_setting("owner_id")
    ready_for_group = await db.get_setting("ready_for_group")

    if str(event.from_user.id) != owner_id:
        await bot.send_message(event.chat.id, "I can only be added by my owner. Leaving...")
        await bot.leave_chat(event.chat.id)
        return

    if ready_for_group != "true":
        await bot.send_message(event.chat.id, "Please use 'Setup Group' in private chat before adding me. Leaving...")
        await bot.leave_chat(event.chat.id)
        return

    existing_group = await db.get_setting("group_id")
    if existing_group and existing_group != str(event.chat.id):
        try:
            await bot.send_message(existing_group, "I am being moved to another group by my owner. Goodbye!")
            await bot.leave_chat(existing_group)
        except Exception as e:
            logger.error(f"Failed to leave old group {existing_group}: {e}")

    await db.set_setting("group_id", str(event.chat.id))
    await db.set_setting("ready_for_group", "false")
    
    is_forum = getattr(event.chat, 'is_forum', False)

    bot_member = await bot.get_chat_member(event.chat.id, bot.id)
    await bot.send_message(
        event.chat.id,
        "Hello! I am ready to manage Jules API tasks here.",
        reply_markup=get_admin_check_keyboard(bot_member, is_forum=is_forum)
    )

@router.my_chat_member(ChatMemberUpdatedFilter(member_status_changed=IS_MEMBER >> ADMINISTRATOR))
async def bot_promoted(event: ChatMemberUpdated, bot: Bot):
    # Whenever bot status changes to admin, we might want to update some pinned message if any,
    # but for now we just rely on the manual check button.
    pass

@router.callback_query(F.data.startswith("admin_"))
async def admin_check_cb(callback: CallbackQuery, bot: Bot):
    if callback.data == "admin_ok":
        await callback.answer("Bot is admin with all required permissions!", show_alert=True)
    elif callback.data == "admin_not_admin":
        # Check current status
        bot_member = await bot.get_chat_member(callback.message.chat.id, bot.id)
        is_forum = getattr(callback.message.chat, 'is_forum', False)
        if bot_member.status == "administrator":
            await callback.message.edit_reply_markup(reply_markup=get_admin_check_keyboard(bot_member, is_forum=is_forum))
            await callback.answer("Status updated!")
        else:
            await callback.answer("Please make the bot an Admin.", show_alert=True)
    elif callback.data == "admin_missing_perms":
        bot_member = await bot.get_chat_member(callback.message.chat.id, bot.id)
        can_manage_topics = getattr(bot_member, 'can_manage_topics', False)
        can_pin_messages = getattr(bot_member, 'can_pin_messages', False)
        is_forum = getattr(callback.message.chat, 'is_forum', False)

        if can_manage_topics and can_pin_messages:
            await callback.message.edit_reply_markup(reply_markup=get_admin_check_keyboard(bot_member, is_forum=is_forum))
            await callback.answer("Permissions updated!")
        else:
            missing = []
            if not can_manage_topics: missing.append("Manage Topics")
            if not can_pin_messages: missing.append("Pin Messages")
            await callback.answer(f"Missing permissions:\n" + "\n".join(missing), show_alert=True)

@router.callback_query(F.data == "forum_not_enabled")
async def forum_not_enabled_cb(callback: CallbackQuery, bot: Bot):
    chat = await bot.get_chat(callback.message.chat.id)
    is_forum = getattr(chat, 'is_forum', False)
    if is_forum:
        bot_member = await bot.get_chat_member(callback.message.chat.id, bot.id)
        await callback.message.edit_reply_markup(reply_markup=get_admin_check_keyboard(bot_member, is_forum=is_forum))
        await callback.answer("Topics are now enabled!", show_alert=True)
    else:
        await callback.answer("Please go to Group Settings -> Enable Topics.", show_alert=True)

File: src/handlers/group.ts
import { Update, ChatMember, InlineKeyboardMarkup, ChatMemberUpdated } from '@grammyjs/types';
import { router } from '../router';
import { DB } from '../db';
import { callApi } from '../api';

function getAdminCheckKeyboard(botMember: ChatMember, isForum: boolean): InlineKeyboardMarkup | undefined {
  const isAdmin = botMember.status === "administrator";
  const canManageTopics = isAdmin && (botMember as any).can_manage_topics === true;
  const canPinMessages = isAdmin && (botMember as any).can_pin_messages === true;

  const keyboard: any[][] = [];

  if (!(isAdmin && canManageTopics && canPinMessages)) {
    if (isAdmin) {
      keyboard.push([{ text: "Permissions missing ⚠️", callback_data: "admin_missing_perms" }]);
    } else {
      keyboard.push([{ text: "Bot is not Admin 🚫", callback_data: "admin_not_admin" }]);
    }
  }

  if (!isForum) {
    keyboard.push([{ text: "Topics not enabled 🚫", callback_data: "forum_not_enabled" }]);
  }

  if (keyboard.length === 0) {
    return undefined;
  }

  return { inline_keyboard: keyboard };
}

// Handler for my_chat_member updates
router.on((u: Update) => !!u.my_chat_member, async (update: Update) => {
  const event = update.my_chat_member!;
  
  // IS_NOT_MEMBER >> IS_MEMBER
  const wasNotMember = ['left', 'kicked', 'restricted'].includes(event.old_chat_member.status) || 
                       (event.old_chat_member.status === 'member' && (event.old_chat_member as any).is_member === false);
  const isMember = event.new_chat_member.status === 'member' || event.new_chat_member.status === 'administrator';

  if (wasNotMember && isMember) {
    if (!['group', 'supergroup'].includes(event.chat.type)) return;

    if (event.chat.type !== 'supergroup') {
      await callApi('sendMessage', { chat_id: event.chat.id, text: "this group is not a supergroup. make it supergroup by enabling topics then add me again" });
      await callApi('leaveChat', { chat_id: event.chat.id });
      return;
    }

    const ownerId = DB.getSetting("owner_id");
    const readyForGroup = DB.getSetting("ready_for_group");

    if (String(event.from.id) !== ownerId) {
      await callApi('sendMessage', { chat_id: event.chat.id, text: "I can only be added by my owner. Leaving..." });
      await callApi('leaveChat', { chat_id: event.chat.id });
      return;
    }

    if (readyForGroup !== "true") {
      await callApi('sendMessage', { chat_id: event.chat.id, text: "Please use 'Setup Group' in private chat before adding me. Leaving..." });
      await callApi('leaveChat', { chat_id: event.chat.id });
      return;
    }

    const existingGroup = DB.getSetting("group_id");
    if (existingGroup && existingGroup !== String(event.chat.id)) {
      try {
        await callApi('sendMessage', { chat_id: existingGroup, text: "I am being moved to another group by my owner. Goodbye!" });
        await callApi('leaveChat', { chat_id: existingGroup });
      } catch (e) {
        console.error(`Failed to leave old group ${existingGroup}:`, e);
      }
    }

    DB.setSetting("group_id", String(event.chat.id));
    DB.setSetting("ready_for_group", "false");

    const isForum = (event.chat as any).is_forum || false;
    const botMemberResponse = await callApi<{ status: string }>('getChatMember', { chat_id: event.chat.id, user_id: event.new_chat_member.user.id });
    
    await callApi('sendMessage', {
      chat_id: event.chat.id,
      text: "Hello! I am ready to manage Jules API tasks here.",
      reply_markup: getAdminCheckKeyboard(botMemberResponse as ChatMember, isForum)
    });
  }
});

router.onCallback('admin_', async (cb, state) => {
  if (!cb.message) return;
  const botInfo = await callApi<any>('getMe');
  const botId = botInfo.id;

  if (cb.data === "admin_ok") {
    await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Bot is admin with all required permissions!", show_alert: true });
  } else if (cb.data === "admin_not_admin") {
    const botMember = await callApi<ChatMember>('getChatMember', { chat_id: cb.message.chat.id, user_id: botId });
    const isForum = (cb.message.chat as any).is_forum || false;

    if (botMember.status === "administrator") {
      try {
        await callApi('editMessageReplyMarkup', {
          chat_id: cb.message.chat.id,
          message_id: cb.message.message_id,
          reply_markup: getAdminCheckKeyboard(botMember, isForum)
        });
      } catch (e) {}
      await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Status updated!" });
    } else {
      await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Please make the bot an Admin.", show_alert: true });
    }
  } else if (cb.data === "admin_missing_perms") {
    const botMember = await callApi<ChatMember>('getChatMember', { chat_id: cb.message.chat.id, user_id: botId });
    const canManageTopics = (botMember as any).can_manage_topics || false;
    const canPinMessages = (botMember as any).can_pin_messages || false;
    const isForum = (cb.message.chat as any).is_forum || false;

    if (canManageTopics && canPinMessages) {
      try {
        await callApi('editMessageReplyMarkup', {
          chat_id: cb.message.chat.id,
          message_id: cb.message.message_id,
          reply_markup: getAdminCheckKeyboard(botMember, isForum)
        });
      } catch(e) {}
      await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Permissions updated!" });
    } else {
      const missing = [];
      if (!canManageTopics) missing.push("Manage Topics");
      if (!canPinMessages) missing.push("Pin Messages");
      await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: `Missing permissions:\n${missing.join('\n')}`, show_alert: true });
    }
  }
});

router.onCallback('forum_not_enabled', async (cb, state) => {
  if (!cb.message) return;
  const botInfo = await callApi<any>('getMe');
  const chat = await callApi<any>('getChat', { chat_id: cb.message.chat.id });
  const isForum = chat.is_forum || false;

  if (isForum) {
    const botMember = await callApi<ChatMember>('getChatMember', { chat_id: cb.message.chat.id, user_id: botInfo.id });
    try {
      await callApi('editMessageReplyMarkup', {
        chat_id: cb.message.chat.id,
        message_id: cb.message.message_id,
        reply_markup: getAdminCheckKeyboard(botMember, isForum)
      });
    } catch(e) {}
    await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Topics are now enabled!", show_alert: true });
  } else {
    await callApi('answerCallbackQuery', { callback_query_id: cb.id, text: "Please go to Group Settings -> Enable Topics.", show_alert: true });
  }
});
